BJS2-78149 Like notification consumer

dextomi opened the pull request.

dextomi committed:
Listener done without logic

dextomi committed:
Listener done without logic

dextomi committed:
Partial changes on notification send

dextomi committed:
Partial changes on notification send

dextomi committed:
Merge remote-tracking branch 'origin/feature/BJS2-78149-like-notification-consumer' into feature/BJS2-78149-like-notification-consumer

# Conflicts:
#	src/main/java/faang/school/notificationservice/config/kafka/KafkaConsumerConfig.java
#	src/main/resources/application.yaml
#	src/main/resources/messages.properties

dextomi committed:
consumer done(finally)

dextomi committed:
removed liquibase debug logging

tanufa25 commented on build.gradle.kts:
```diff
@@ -24,12 +24,13 @@ dependencies {
     implementation("org.springframework.boot:spring-boot-starter-mail:3.0.6")
     implementation("org.springframework.cloud:spring-cloud-starter-openfeign:4.0.2")
-    implementation ("org.springframework.kafka:spring-kafka")
+    implementation("org.springframework.kafka:spring-kafka")
```
Тут в 52 строке тоже кафка есть уже

dextomi replied:
Поправил и удалил все дубликаты зависимостей кафки

tanufa25 commented on src/main/java/faang/school/notificationservice/messaging/like/CommentLikedEventMessageBuilder.java:
```diff
@@ -0,0 +1,77 @@
+@RequiredArgsConstructor
+public class CommentLikedEventMessageBuilder implements MessageBuilder<AggregatedNotificationsDto> {
+
+    private static final int MAX_MESSAGE_LENGTH = 100;
```
Лучше вынести в настройки

tanufa25 commented on src/main/java/faang/school/notificationservice/messaging/like/PostLikedEventMessageBuilder.java:
```diff
@@ -0,0 +1,78 @@
+@RequiredArgsConstructor
+public class PostLikedEventMessageBuilder implements MessageBuilder<AggregatedNotificationsDto> {
+
+    private static final int MAX_MESSAGE_LENGTH = 100;
```
Вынести в application.yaml

tanufa25 commented on src/main/java/faang/school/notificationservice/service/notification/NotificationSenderService.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    private void sendMessageByType(AggregatedNotificationsDto notification, EventType eventType) {
+        UserDto user = userServiceClient.getUser(notification.getReceiverId());
```
У нас же в кафке приходит UserDTO owner, может лучше сразу в AggregatedNotificationsDto положить его, а не просто id, а то как-то слишком энергозатратно в UserService ходить.

tanufa25 replied:
Поняла, он же в базу у тебя кладётся, и из неё запросом достаются значения.
Но как-то слишком сложно в плане межсервисного взаимодействия получается.
Понимаю, что это не предусмотрено задачей, и так слишком много всего получилось.
Но мысли вслух по этой теме. 
Раз уж нам по кафке и так приходит инфа про юзеров-получателей, а положить целого юзера в базу сложно в рамках другого микросервиса (да и не нужно), то, может, в кеш отправлять юзеров-отправителей с ttl чуть больше, чем период сбора агрегированных сообщений, чтобы не бегать каждый раз в микросервис юзеров, и там из базы доставать? А если вдруг из кеша инфа пропадёт, тогда уж сходить к user сервису.

dextomi replied:
Мысль была в том, чтобы брать актуальную информацию о юзере, ведь в БД она всегда будет такой, мало ли, может контакт какой-то поменяется в моменте, а в кэше будет другой и пользаку уведомление не придет

20cognosce replied:
Не надо долбить другие сервисы рестом, всё должно идти только по кафке

tanufa25 commented on src/main/resources/application.yaml:
```diff
@@ -1,6 +1,9 @@
 spring:
   kafka:
     bootstrap-servers: localhost:9092
+    producer:
+      key-serializer: org.apache.kafka.common.serialization.StringSerializer
+      value-serializer: org.springframework.kafka.support.serializer.JsonSerializer
```
А зачем здесь настройки продюсера? Этот микросервис пока только как консьюмер, и его настройки прописаны в KafkaConsumerConfig.

dextomi replied:
Почему то в DLQ не падали сообщения из-за того, что по умолчанию используется StringSerializer для значений, вот я и по умолчанию здесь прописал Json

tanufa25 commented on src/main/java/faang/school/notificationservice/listener/AbstractEventListener.java:
```diff
@@ -2,51 +2,15 @@
 
 import java.util.Arrays;
-import java.util.List;
-import java.util.Locale;
 import java.util.Objects;
 import java.util.function.Supplier;
 
-@Slf4j
-@RequiredArgsConstructor
 public abstract class AbstractEventListener<T extends NotificationEvent> {
 
-    private final List<NotificationService> notificationServices;
-    private final MessageBuilder<T> messageBuilder;
-
     public abstract boolean isEventValid(T event);
 
-    protected void sendNotification(T event) {
-        if (isEventValid(event)) {
-            sendMessage(event.getOwner(), getMessage(event));
-        } else {
-            log.error("Event validation failed. Event: {}", event);
-        }
-    }
-
-    protected String getMessage(T event) {
-        Locale locale = Objects.isNull(event.getOwner().getLocale())
-                ? LocaleContextHolder.getLocale()
-                : event.getOwner().getLocale();
-        return messageBuilder.buildMessage(event, locale);
-    }
-
-    protected void sendMessage(UserDto userDto, String message) {
-        notificationServices.stream()
-                .filter(notificationService -> notificationService.getPreferredContact() == userDto.getPreference())
-                .findFirst()
-                .orElseThrow(() -> new IllegalArgumentException(
-                        "No notification service was found for user preferred notification type"))
-                .send(userDto, message);
-    }
-
```
А можешь, пожалуйста, пояснить, почему ты расформировал методы AbstractEventListener в DirectNotificationEventListener?

dextomi replied:
В текущей реализации все слушатели сразу отправляют уведомления пользователю, а у меня сохраняются в БД.  Чтобы все не перекапывать я подумал так сделать и назвал его так, не знаю правда, насколько это правильно с точки зрения структуры

tanufa25 commented on src/main/java/faang/school/notificationservice/messaging/like/CommentLikedEventMessageBuilder.java:
```diff
@@ -0,0 +1,77 @@
+        return userServiceClient.getUser(userId);
+    }
+
+    private LikeDto getLike(Long likeId) {
+        return postServiceClient.getLike(likeId);
+    }
+
+    private String getCommentContent(Long commentId) {
+        return postServiceClient.getComment(commentId).getContent();
+    }
```
Как мне кажется, тут тоже лучше в kafka сразу короткий контент передавать, и его либо в базу в отдельную колонку PendingNotifications, либо в кеш добавлять. И для лайка нужен же только username, тоже лучше его в кафке сразу передавать, и либо тоже в новую колонку PendingNotifications, либо в кеш, чем постоянно туда-сюда между микросервисами запросы гонять.

dextomi replied:
У меня pending_notifications получилась универсальной, из-за чего я не мог какое-то определенное решение придумать, поэтому получилось так. У меня нет идей как это можно сделать по другому

dextomi committed:
Deleted kafka dependency duplicates

dextomi committed:
Moved some variable value into config file

dextomi committed:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-78149-like-notification-consumer

# Conflicts:
#	src/main/resources/application.yaml
#	src/main/resources/db/changelog/db.changelog-master.yaml

dextomi committed:
Added old notifications cleanup and made scheduler thread safe

dextomi committed:
fixed some unit tests

dextomi committed:
fixed unit tests and wrote tests on my services

20cognosce commented on src/main/java/faang/school/notificationservice/client/PostServiceClient.java:
```diff
@@ -0,0 +1,22 @@
+@FeignClient(name = "post-service", url = "${post-service.host}:${post-service.port}")
+public interface PostServiceClient {
+
+    @GetMapping("api/v1/posts")
+    PostDto getPost(@RequestParam long postId);
+
+    @GetMapping("api/v1/comments")
+    CommentDto getComment(@RequestParam long commentId);
+
+    @GetMapping("api/v1/like/{likeId}")
+    LikeDto getLike(@PathVariable long likeId);
```
Нет, нет, нет
Этого быть не должно, всё приходит по кафке

20cognosce commented on src/main/java/faang/school/notificationservice/service/notification/handler/PostLikedNotificationEventHandler.java:
```diff
@@ -0,0 +1,38 @@
+                        PendingNotificationsDto.builder()
+                                .receiverId(event.getOwner().getId())
+                                .targetEntityId(event.getPostId())
+                                .relatedEntityId(event.getLikeId())
```
а зачем две сущности? окей, тут видимо id в сообщениях в кафке тебе нужны, но только для группировки по буферам

но группируешь ты только по одному посту или одному комменту, тебе одной сущности достаточно

20cognosce commented on src/main/resources/db/changelog/changeset/notification_V001__initial.sql:
```diff
@@ -0,0 +1,21 @@
+    id                BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+    receiver_id       BIGINT      NOT NULL,
+    target_entity_id  BIGINT      NOT NULL,
+    related_entity_id BIGINT,
```
id одной сущности для группировки достаточно

20cognosce replied:
related_entity - непонятное поле

20cognosce commented on src/main/java/faang/school/notificationservice/messaging/like/PostLikedEventMessageBuilder.java:
```diff
@@ -0,0 +1,80 @@
+
+    private String getPostContent(Long postId) {
+        return postServiceClient.getPost(postId).getContent();
+    }
```
это плохо, лишняя нагрузка

20cognosce commented on src/main/java/faang/school/notificationservice/messaging/like/CommentLikedEventMessageBuilder.java:
```diff
@@ -0,0 +1,79 @@
+        String messageTemplateCode = "comment.liked";
+        String content = getCommentContent(commentId);
+
+        return messageSource.getMessage(messageTemplateCode,
```
у тебя для комментариев и постов во многих местах код дублируется

20cognosce commented on src/main/java/faang/school/notificationservice/listener/like/PostLikedEventListener.java:
```diff
@@ -0,0 +1,32 @@
+            containerFactory = "kafkaPostLikedEventListener"
+    )
+    public void listenPostLikedBatch(List<PostLikedNotificationEvent> events) {
+        eventHandler.saveNotifications(events);
```
а почему ты тут просто handler вызываешь, а не processBatch?

20cognosce commented on src/main/java/faang/school/notificationservice/scheduler/OldNotificationsCleanupJob.java:
```diff
@@ -0,0 +1,36 @@
+    @Value("${notification-cleanup.cleanup-interval}")
+    private int cleanupInterval;
+
+    @Scheduled(cron = "${notification-cleanup.cron}")
+    @Transactional
```
вместе не использовать два аспекта на одном методе

20cognosce commented on src/main/java/faang/school/notificationservice/scheduler/NotificationSchedulerJob.java:
```diff
@@ -0,0 +1,57 @@
+    @Value("${notification-scheduler.maxRetryAttempts}")
+    private int maxRetryAttempts;
+
+    @Scheduled(cron = "${notification-scheduler.cron}")
+    @Transactional
```
.

20cognosce commented on src/main/java/faang/school/notificationservice/repository/NotificationRepository.java:
```diff
@@ -0,0 +1,63 @@
+                    SELECT *
+                    FROM pending_notifications pn
+                    WHERE pn.status IN ('PENDING', 'FAILED')
+                    AND retry_count < :maxRetryAttempts
```
мне кажется запись о ретраях не должна хранится в БД
ретрай сразу, а не по крону джобы - если не получилось, то всё значит

очень громоздко

20cognosce commented on src/main/java/faang/school/notificationservice/repository/NotificationRepository.java:
```diff
@@ -0,0 +1,63 @@
+                            AND sub.sent_at >= :lastSentThreshold
+                      )
+                    FOR UPDATE SKIP LOCKED
+                    LIMIT 10000
```
А что, если ты в этом лимите захватил не все уведомления
10001 лайк пришел по одному посту

20cognosce replied:
И как несколько инстансов будут их делить? правильнее ведь, чтобы один инстанс полностью обрабатывал все события сгруппированные по одной конкретной сущности

20cognosce commented on src/main/java/faang/school/notificationservice/repository/NotificationRepository.java:
```diff
@@ -0,0 +1,63 @@
+                    WHERE pn.status IN ('PENDING', 'FAILED')
+                    AND retry_count < :maxRetryAttempts
+                      AND pn.created_at <= :notificationDelay
+                      AND NOT EXISTS (
+                          SELECT 1
+                          FROM pending_notifications sub
+                          WHERE sub.status = 'SENT'
+                            AND sub.target_entity_id = pn.target_entity_id
+                            AND sub.sent_at >= :lastSentThreshold
```
сравнивать с другими уведомлениями конечно можно, но это крайне неэффективно ведь
не проще по receiver_id сделать еще одну таблицу с датами последних уведомлений?

20cognosce reviewed (changes requested):
Неплохо, но по уровню БД и рестовым запросам есть сомнения

dextomi committed:
some fixes in notification repository;
created new table;
made abstract class for comment and like message builders to remove duplicated code;
mb some minor fixes i don't remember about already)
added @Transactional on updateGroup method in repository, it's only used in one place and shoudn't harm anything

20cognosce commented on src/main/java/faang/school/notificationservice/dto/notification/AggregatedNotificationsDto.java:
```diff
@@ -0,0 +1,20 @@
+public class AggregatedNotificationsDto {
+    private Long receiverId;
+    private Long targetEntityId;
+    private JsonNode eventData;
```
в JsonNode может быть любая структура данных, никакой DTO не получилось создать?

20cognosce commented on src/main/java/faang/school/notificationservice/service/notification/handler/CommentLikedNotificationEventHandler.java:
```diff
@@ -0,0 +1,40 @@
+    private final ObjectMapper objectMapper;
+
+    @Override
+    @Transactional
```
зачем

20cognosce commented on src/main/java/faang/school/notificationservice/scheduler/NotificationSchedulerJob.java:
```diff
@@ -0,0 +1,63 @@
+        LocalDateTime notificationDelay = LocalDateTime.now().minusHours(delayHours);
+        LocalDateTime lastSentThreshold = LocalDateTime.now().minusHours(lastSentThresholdHours);
+
+        List<PendingNotifications> pendingNotifications = notificationRepository
+                .findAndLockPendingNotifications(
+                        notificationDelay,
+                        lastSentThreshold,
+                        totalInstances,
+                        currentInstance
+                );
```
локать нужно внутри транзакции, сейчас это бесполезно - следующей же строчкой лок снимется

20cognosce commented on src/main/java/faang/school/notificationservice/service/notification/NotificationSenderService.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            sendMessage(notification, eventType);
+
+            notificationRepository.updateStatusByGroup(
+                    notification.getReceiverId(),
+                    notification.getTargetEntityId(),
+                    eventType.name(),
+                    NotificationStatus.SENT.name()
+            );
```
почему ты по DTO выставляешь статус, по сущности PendingNotifications (её id) нельзя статус проставлять?

20cognosce commented on src/main/java/faang/school/notificationservice/repository/NotificationRepository.java:
```diff
@@ -0,0 +1,78 @@
+    @Query(
+            nativeQuery = true,
+            value = """
+                        WITH candidate_ids AS (
+                            SELECT
+                                MIN(pn.id) AS id
+                            FROM pending_notifications pn
+                            LEFT JOIN last_sent_notifications lsn
+                              ON lsn.receiver_id = pn.receiver_id
+                             AND lsn.target_entity_id = pn.target_entity_id
+                             AND lsn.event_type = pn.event_type
+                            WHERE
+                                pn.status = 'PENDING'
+                                AND pn.created_at <= :notificationDelay
+                                AND (lsn.last_sent_at IS NULL OR lsn.last_sent_at < :lastSentThreshold)
+                                AND (
+                                    :totalInstances = 1
+                                    OR MOD(HASHTEXT(CONCAT(pn.target_entity_id, ':', pn.event_type)), :totalInstances) = :currentInstance
+                                )
+                            GROUP BY pn.receiver_id, pn.target_entity_id, pn.event_type
+                            ORDER BY MIN(pn.created_at)
+                            LIMIT 1000
+                        )
+                        SELECT pn.*
+                        FROM pending_notifications pn
+                        JOIN candidate_ids ci ON pn.id = ci.id
+                        FOR UPDATE SKIP LOCKED
```
зачем эта жесть
это невозможно поддерживать

декомпозировать

20cognosce reviewed (dismissed):
Принято с замечаниями
